src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -0,0 +1,574 @@
+package dotty.tools.dotc
+package transform
+package patmat
+
+import core.Types._
+import core.Contexts._
+import core.Flags._
+import ast.Trees._
+import ast.tpd
+import core.Decorators._
+import core.TypeApplications._
+import core.Symbols._
+import core.StdNames._
+import core.NameOps._
+import core.Constants._
+
+/** Space logic for checking exhaustivity and unreachability of pattern matching
+ *
+ *  Space can be thought of as a set of possible values. A type or a pattern
+ *  both refer to spaces. The space of a type is the values that inhabit the
+ *  type. The space of a pattern is the values that can be covered by the
+ *  pattern.
+ *
+ *  Space is recursively defined as follows:
+ *
+ *      1. `Empty` is a space
+ *      2. For a type T, `Typ(T)` is a space
+ *      3. A union of spaces `S1 | S2 | ...` is a space
+ *      4. For a case class Kon(x1: T1, x2: T2, .., xn: Tn), if S1, S2, ..., Sn
+ *          are spaces, then `Kon(S1, S2, ..., Sn)` is a space.
+ *      5. A constant `Const(value, T)` is a point in space
+ *      6. A stable identifier `Var(sym, T)` is a space
+ *
+ *  For the problem of exhaustivity check, its formulation in terms of space is as follows:
+ *
+ *      Is the space Typ(T) a subspace of the union of space covered by all the patterns?
+ *
+ *  The problem of unreachable patterns can be formulated as follows:
+ *
+ *      Is the space covered by a pattern a subspace of the space covered by previous patterns?
+ */
+
+
+/** space definition */
+sealed trait Space
+case object Empty extends Space
+case class Typ(tp: Type, decomposed: Boolean) extends Space
+case class Kon(tp: Type, params: List[Space]) extends Space
+sealed trait Point extends Space
+case class Var(sym: Symbol, tp: Type) extends Point
+case class Const(value: Constant, tp: Type) extends Point
+case class Or(spaces: List[Space]) extends Space
+
+/** abstract space logic */
+trait SpaceLogic {
+  /** Is `tp1` a subtype of `tp2`? */
+  def isSubType(tp1: Type, tp2: Type): Boolean
```

Comment:
there may be huge cans of worms here as you are using pre-erasure `<:<` on semi-erazed or fully-erazed types.
Maybe it would make sense to go into full erasure and replace sub typing by `derivesFrom` after deconstructing Or types into your OrSpaces.